source/gloo_proxy.go

```diff
@@ -93,6 +117,32 @@ func NewGlooSource(dynamicKubeClient dynamic.Interface, kubeClient kubernetes.In
 	}, nil
 }
 
+// virtualHosts aggregates all virtual hosts from the available listener types.
+// Only 1 of HybridListener or HTTPListener will be set
+func (l proxySpecListener) virtualHosts() []proxyVirtualHost {
+	virtualHosts := l.HTTPListener.VirtualHosts
+	for _, hybridMatchedListener := range l.HybridListener.MatchedListeners {
+		virtualHosts = append(virtualHosts, hybridMatchedListener.HTTPListener.VirtualHosts...)
+	}
+	return virtualHosts
+}
+
+// sources aggregates all sources within a proxyVirtualHost.
+// Only 1 of Metadata for MetadataStatic will be set
+func (vh proxyVirtualHost) sources() []proxyVirtualHostMetadataSource {
+	sources := vh.Metadata.Source
+	for _, srcStatic := range vh.MetadataStatic.Source {
```

Comment:
same question about nil safety here between `vh.Metadata` and `vh.MetadataStatic`

Reply:
nevermind, i see these are structs not pointers to structs